### PR TITLE
Typeclasses default mode

### PR DESCRIPTION
**Pull request by mattam82**
- Allow to set globally a default mode for all parameters of a class
- Add an attribute that overrides this on a class by class basis
- Add a settable warning for using the default mode
- The default is still "-", meaning no mode filtering and full compatibility.

- [x] Added / updated **test-suite**.

- [x] Added **changelog**.
- [x] Added / updated **documentation**.
  
  - [x] Documented any new / changed **user messages**.

- [x] Opened **overlay** pull requests.
https://github.com/LPCIC/coq-elpi/pull/683

### CODE REVIEW

**Comment by mattam82** [comment]
@coqbot full ci

**Comment by Alizter** [comment]
One thing that is missing in the `Hint Mode` documentation is an example of what those settings mean in practice. This should probably be improved in a later PR with some examples in the documentaiton.

What does the change of default hint mode mean here in practice for typeclasses?

**Comment by mattam82** [comment]
I backtracked on changing the default, too much incompatibilities arise. If one changes it from "-" to "!" or "+" it will generally incur more failed typeclass searches (or speedups in successful searches). I can add more doc, that's really necessary.

File: doc/sphinx/addendum/type-classes.rst [hunk 1]
```diff
@@ -564,6 +573,28 @@ type, like:
...
+.. opt:: Typeclasses Default Mode {| "+" | "-" | "!" }.
+
+   Sets the default mode declaration associated with a :cmd:`Class` or :cmd:`Existing Class`
+   declaration. In Coq <= 8.20, it was set by default to "-", i.e. doing no mode filtering
+   by default. From Coq 8.21 on, it is set to "!", meaning each parameter should be headed by
+   something else than an existential variable for resolution to proceed. If the option is set,
+   then each class declaration uses this default mode for *all* its parameters, unless a
+   :attr:`mode` attribute is used to set the mode explicitly.
+
+   .. _TypeclassesDefaultModeWarning:
```
Comment (jfehrle) [inline doc/sphinx/addendum/type-classes.rst:587]:
```suggestion
```

Label is not used, can be deleted.  If you want to keep it, you might use the warning name `class-declaration-default-mode` as the label.  Maybe that would be a good convention.  (Or someday we could enhance `.. warn::` to include the warning name in a standard way.)

**Review by jfehrle — COMMENTED** [review]
wording suggestions

File: doc/sphinx/addendum/type-classes.rst [hunk 2]
```diff
@@ -564,6 +573,28 @@ type, like:
...
+.. _TypeclassesDefaultMode:
+
+.. opt:: Typeclasses Default Mode {| "+" | "-" | "!" }.
+
+   Sets the default mode declaration associated with a :cmd:`Class` or :cmd:`Existing Class`
+   declaration. In Coq <= 8.20, it was set by default to "-", i.e. doing no mode filtering
+   by default. From Coq 8.21 on, it is set to "!", meaning each parameter should be headed by
+   something else than an existential variable for resolution to proceed. If the option is set,
+   then each class declaration uses this default mode for *all* its parameters, unless a
+   :attr:`mode` attribute is used to set the mode explicitly.
```
[the remark this answers is not in the document]
Comment (jfehrle) [reply doc/sphinx/addendum/type-classes.rst:585]: Would it ever make sense to unset the mode entirely?  If so, how does the user do that?

**Comment by TheoWinterhalter** [comment]
Instead of changing the global default for everyone, my suggestion would be to do it for new user-created stuff, like proposed in #19117.

**Comment by ppedrot** [comment]
Indeed, the change of default hint locality is a somewhat successful prior of this technique. We could first warn when modes are missing, turn this at some point in an error for one version and then switch the default.

**Comment by mattam82** [comment]
Agreed @TheoWinterhalter, that was the original motivation of this PR to be able to set the default mode to "!" in GoodDefaults :)

**Comment by mattam82** [comment]
@coqbot run full ci

File: doc/sphinx/addendum/type-classes.rst [hunk 3]
```diff
@@ -317,13 +317,22 @@ Command summary
...
+   declaration for the class, see :cmd:`Hint Mode`.
+
    .. note::
       Don't confuse typeclasses with "coercion classes", described in
       `implicit coercions<classes-implicit-coercions>`.
 
    When record syntax is used, this command also supports the
    :attr:`projections(primitive)` :term:`attribute`.
 
+   .. attr:: mode
```
[the remark this answers is not in the document]
Comment (mattam82) [reply doc/sphinx/addendum/type-classes.rst:330]: In the context of a class declaration (which is the only command supporting this attribute) it makes sense to use "mode".

File: doc/sphinx/addendum/type-classes.rst [hunk 4]
```diff
@@ -564,6 +573,28 @@ type, like:
...
 ~~~~~~~~
 
+.. _TypeclassesDefaultMode:
+
+.. opt:: Typeclasses Default Mode {| "+" | "-" | "!" }.
+
+   Sets the default mode declaration associated with a :cmd:`Class` or :cmd:`Existing Class`
+   declaration. In Coq <= 8.20, it was set by default to "-", i.e. doing no mode filtering
+   by default. From Coq 8.21 on, it is set to "!", meaning each parameter should be headed by
+   something else than an existential variable for resolution to proceed. If the option is set,
```
[the remark this answers is not in the document]
Comment (mattam82) [reply doc/sphinx/addendum/type-classes.rst:583]: Done

File: doc/sphinx/addendum/type-classes.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (mattam82) [reply doc/sphinx/addendum/type-classes.rst:585]: Nope, it doesn't make sense. The "-" mode for all parameters corresponds to the "unset" setting already.

File: doc/sphinx/addendum/type-classes.rst (hunk 1 above)
Comment (mattam82) [reply doc/sphinx/addendum/type-classes.rst:587]: It will be useful in the changelog entry.

File: doc/sphinx/addendum/type-classes.rst [hunk 5]
```diff
@@ -564,6 +573,28 @@ type, like:
...
+   .. _TypeclassesDefaultModeWarning:
+
+   .. warn:: Using inferred default mode: “mode” for “@ident”
+
+      Indicates that the :attr:`mode` for a :cmd:`Class` declaration have been
+      assigned automatically using the default mode.
+      This warning is named ``class-declaration-default-mode``.
+      It is disabled by default, see :opt:`Warnings` to enable it or turn it into an error.
+      It can be used to lint a library and ensure all typeclasses have been assigned
+      explicit mode declarations.
```
[the remark this answers is not in the document]
Comment (mattam82) [reply doc/sphinx/addendum/type-classes.rst:596]: Done

**Comment by mattam82** [comment]
I backtracked on trying to change the default or any mode declaration in the stdlib to ensure compatibility for now.

**Comment by mattam82** [comment]
@coqbot run full ci

File: vernac/vernacentries.ml [hunk 6]
```diff
@@ -882,16 +900,21 @@ let preprocess_inductive_decl ~atts kind indl =
...
+    | Class _ -> mode_attr
+    | _ -> Notations.return None
+  in
+  let ((((template, (poly, cumulative)), private_ind), typing_flags), primitive_proj), mode =
     Attributes.(
       parse Notations.(
           template
           ++ polymorphic_cumulative ~is_defclass:(Option.has_some is_defclass)
-          ++ private_ind ++ typing_flags ++ prim_proj_attr)
+          ++ private_ind ++ typing_flags ++ prim_proj_attr ++ hint_mode_attr)
```
Comment (gares) [inline vernac/vernacentries.ml:913]: This attribute does not seem to be tested in the test suite and also it does not error (just ignore it) if the inductive is not a Class.

**Comment by gares** [comment]
CC @FissoreD , this PR allows one to declare modes together with the class

File: vernac/vernacentries.ml (hunk 6 above)
Comment (mattam82) [reply vernac/vernacentries.ml:916]: Now tested. It does fail if the command is not a Class.

**Comment by mattam82** [comment]
@coqbot run full ci

**Comment by mattam82** [comment]
It just needs a changelog and review now.

**Comment by mattam82** [comment]
@gares could you assign and merge?

File: vernac/vernacentries.ml [hunk 7]
```diff
@@ -842,6 +842,24 @@ let primitive_proj =
 let { Goptions.get = do_auto_prop_lowering } =
   Goptions.declare_bool_option_and_ref ~key:["Automatic";"Proposition";"Inductives"] ~value:true ()
 
+let parse_mode s =
+  match s with
+  | "+" -> Hints.ModeInput
+  | "-" -> Hints.ModeOutput
+  | "!" -> Hints.ModeNoHeadEvar
+  | _ -> CErrors.user_err Pp.(str"Unrecognized hint mode " ++ str s)
```
Comment (SkySkimmer) [inline vernac/vernacentries.ml:850]: Duplicated from code in record.ml

File: pretyping/typeclasses.ml [hunk 8]
```diff
@@ -263,12 +263,14 @@ let resolve_typeclasses ?(filter=no_goals) ?(unique=get_typeclasses_unique_solut
...
-    if Option.has_some accu then raise MultipleFound
-    else (Some ev)
+    match Evd.find_undefined evd ev with
+    | exception Not_found -> None
+    | evi ->
+      let ev_class = class_of_constr env evd (Evd.evar_concl evi) in
+      if Option.is_empty ev_class then accu
+      else (* focus on one instance if only one was searched for *)
+      if Option.has_some accu then raise MultipleFound
+      else (Some ev)
```
Comment (SkySkimmer) [inline pretyping/typeclasses.ml:273]: why this change?

File: tactics/hints.ml [hunk 9]
```diff
@@ -1007,8 +1007,8 @@ let make_extern pri pat tacast =
 let make_mode ref m =
   let open Term in
   let ty, _ = Typeops.type_of_global_in_context (Global.env ()) ref in
-  let ctx, t = decompose_prod ty in
-  let n = List.length ctx in
+  let ctx, t = decompose_prod_decls ty in
+  let n = Context.Rel.nhyps ctx in
```
Comment (SkySkimmer) [inline tactics/hints.ml:1011]: why this change?

File: pretyping/typeclasses.ml (hunk 8 above)
Comment (mattam82) [reply pretyping/typeclasses.ml:273]: This is a bugfix

File: tactics/hints.ml (hunk 9 above)
Comment (mattam82) [reply tactics/hints.ml:1011]: Another bugfix: let-bindings should not matter for the mode decl

File: vernac/vernacentries.ml (hunk 7 above)
Comment (mattam82) [reply vernac/vernacentries.ml:850]: Right...

File: doc/sphinx/addendum/type-classes.rst [hunk 10]
```diff
@@ -317,13 +317,21 @@ Command summary
...
       Don't confuse typeclasses with "coercion classes", described in
       `implicit coercions<classes-implicit-coercions>`.
 
    When record syntax is used, this command also supports the
    :attr:`projections(primitive)` :term:`attribute`.
 
+   .. attr:: mode({+ {| + | ! | - } })
+      :name: mode
+
+      Sets the mode of resolution for queries on the class. The syntax is explained in :cmd:`Hint Mode`.
```
Comment (gares) [inline doc/sphinx/addendum/type-classes.rst:333]: The syntax seems wrong here, e.g. it is `mode="+ + - !"` not `mode=(+ + - !)`

File: vernac/record.ml [hunk 11]
```diff
@@ -922,6 +940,29 @@ let declare_class_constant ~univs paramimpls params data =
...
+    | _ ->
+      let ctxl = Context.Rel.nhyps ctx in
+      let mode = match typeclasses_default_mode () with
+      | Hints.ModeOutput -> None
+      | Hints.ModeInput ->
+        Some (List.init ctxl (fun _ -> Hints.ModeInput))
+      | Hints.ModeNoHeadEvar ->
+        Some (List.init ctxl (fun _ -> Hints.ModeNoHeadEvar))
+      in
+      let wm = match mode with None -> List.init ctxl (fun _ -> Hints.ModeOutput) | Some m -> m in
```
Comment (gares) [inline vernac/record.ml:957]: nitpick
```suggestion
      let wm =
        let def = typeclasses_default_mode () in
        List.init ctxl (fun _ -> def) in
```

Comment (mattam82) [reply vernac/record.ml:957]: I simplified a bit.

File: doc/sphinx/addendum/type-classes.rst (hunk 10 above)
Comment (mattam82) [reply doc/sphinx/addendum/type-classes.rst]: Indeed, I did the same as is done for the warning attribute and refered to Hint Mode for the syntax.

File: vernac/vernacentries.ml (hunk 7 above)
Comment (mattam82) [reply vernac/vernacentries.ml]: Now moved

File: pretyping/typeclasses.ml (hunk 8 above)
Comment (mattam82) [reply pretyping/typeclasses.ml:273]: It would otherwise result in an uncaught exception

Comment (SkySkimmer) [reply pretyping/typeclasses.ml:273]: Why though? What's the test case?

**Comment by gares** [comment]
@coqbot run full ci

File: pretyping/typeclasses.ml (hunk 8 above)
Comment (gares) [reply pretyping/typeclasses.ml:273]: I guess the evar in question could be solved by a previous TC resolution. Does this make sense?

File: doc/sphinx/addendum/type-classes.rst [hunk 12]
```diff
@@ -564,6 +573,25 @@ type, like:
 Settings
 ~~~~~~~~
 
+.. _TypeclassesDefaultMode:
+
+.. opt:: Typeclasses Default Mode {| "+" | "-" | "!" }.
```
Comment (jfehrle) [inline doc/sphinx/addendum/type-classes.rst:578]:
```suggestion
.. opt:: Typeclasses Default Mode {| "+" | "-" | "!" }
```

**Review by jfehrle — COMMENTED** [review]
Further suggestions

File: pretyping/typeclasses.ml (hunk 8 above)
Comment (mattam82) [reply pretyping/typeclasses.ml:273]: Yes, it could be solved from another resolution in principle.

**Review by gares — APPROVED** [review]

File: vernac/record.ml [hunk 13]
```diff
@@ -40,6 +40,17 @@ let { Goptions.get = typeclasses_unique } =
     ~value:false
     ()
 
+let { Goptions.get = typeclasses_default_mode } =
+  let interp = function
+    | Hints.ModeInput -> "+"
+    | Hints.ModeOutput -> "-"
+    | Hints.ModeNoHeadEvar -> "!"
+  in
```
Comment (gares) [inline vernac/record.ml:48]: I would prefer to have this code next to the one for parsing modes.

**Comment by gares** [comment]
@mattam82 I think we can merge, but I also think you could rebase and squash.
If you want to apply my last nit while doing so fine, otherwise it is not the end of the world.

Just tell me if you do any other change, cause CI was already green and I'd rather not run it again if not needed.

**Comment by mattam82** [comment]
I just did your nitpick and squashed, factorizing with code in hints.ml (string_of_mode). A light pipeline should be enough here.

**Comment by gares** [comment]
@coqbot merge now

**Comment by gares** [comment]
sudo @coqbot merge now
